handler/rfc8628/token_endpoint_handler.go

```diff
@@ -0,0 +1,62 @@
+// Copyright © 2023 Ory Corp
+// SPDX-License-Identifier: Apache-2.0
+
+package rfc8628
+
+import (
+	"context"
+
+	"github.com/ory/fosite/handler/oauth2"
+	"github.com/ory/x/errorsx"
+
+	"github.com/ory/fosite"
+)
+
+// DeviceAuthorizeHandler is a response handler for the Device Code introduced in the Device Authorize Grant
+// as defined in https://www.rfc-editor.org/rfc/rfc8628
+type DeviceAuthorizeHandler struct {
+	DeviceStrategy DeviceCodeStrategy
+	DeviceStorage  DeviceCodeStorage
+}
+
+type DeviceCodeTokenEndpointHandler struct {
+	oauth2.GenericCodeTokenEndpointHandler
+}
+
+var _ oauth2.CodeTokenEndpointHandler = (*DeviceAuthorizeHandler)(nil)
+var _ fosite.TokenEndpointHandler = (*DeviceCodeTokenEndpointHandler)(nil)
+
+func (c *DeviceAuthorizeHandler) ValidateGrantTypes(ctx context.Context, requester fosite.AccessRequester) error {
+	if !requester.GetClient().GetGrantTypes().Has(string(fosite.GrantTypeDeviceCode)) {
+		return errorsx.WithStack(fosite.ErrUnauthorizedClient.WithHint("The OAuth 2.0 Client is not allowed to use authorization grant \"urn:ietf:params:oauth:grant-type:device_code\"."))
+	}
+
+	return nil
+}
+
+func (c *DeviceAuthorizeHandler) ValidateCode(ctx context.Context, request fosite.AccessRequester, code string) error {
+	return c.DeviceStrategy.ValidateDeviceCode(ctx, request, code)
+}
+
+func (c *DeviceAuthorizeHandler) GetCodeAndSession(ctx context.Context, requester fosite.AccessRequester) (code string, signature string, request fosite.Requester, err error) {
+	code = requester.GetRequestForm().Get("device_code")
+	signature, err = c.DeviceStrategy.DeviceCodeSignature(ctx, code)
+	if err != nil {
+		return "", "", nil, errorsx.WithStack(fosite.ErrServerError.WithWrap(err).WithDebug(err.Error()))
+	}
+	req, err := c.DeviceStorage.GetDeviceCodeSession(ctx, signature, requester.GetSession())
+	return code, signature, req, err
+}
+
+func (c *DeviceAuthorizeHandler) InvalidateSession(ctx context.Context, signature string) error {
+	return c.DeviceStorage.InvalidateDeviceCodeSession(ctx, signature)
+}
+
+// implement TokenEndpointHandler
+func (c *DeviceAuthorizeHandler) CanSkipClientAuth(ctx context.Context, requester fosite.AccessRequester) bool {
+	return requester.GetGrantTypes().ExactOne(string(fosite.GrantTypeDeviceCode))
+}
+
+func (c *DeviceAuthorizeHandler) CanHandleTokenEndpointRequest(ctx context.Context, requester fosite.AccessRequester) bool {
+	return requester.GetGrantTypes().ExactOne(string(fosite.GrantTypeDeviceCode))
```

Comment:
Maybe I missed it elsewhere, but where are we checking the rate-limiting?

Reply:
There is no rate-limiting in the code at the moment.

Reply:
I'd put the interface into fosite in this PR, because now we have the most context.

Reply:
I'm not sure how to tacke this. :(
Where should I put this interface? 
How does Hydra handle multi-node configuration ?

Reply:
Hi @hperl,

I've implemented something simple, see here:
https://github.com/BuzzBumbleBee/fosite/blob/feat_dev_grants_2x/handler/rfc8628/strategy_hmacsha.go#L86

Let me know if it fits the requirement. I've tested it and it works quite well.